refactor: Align on gov/group Proposals and Vote syntax

amaurym opened the pull request:
<!--
The default pull request template is for types feat, fix, or refactor.
For other templates, add one of the following parameters to the url:
- template=docs.md
- template=other.md
-->

## Description

Closes: #11070

- Proto-breaking changes:
  - group Choice -> VoteOption
  - group MsgCreateProposal -> MsgSubmitProposal
  - group enums -> prepend `PROPOSAL_` (e.g. `PROPOSAL_STATUS_`, `PROPOSAL_RESULT_`)
  - group Tally -> TallyResult
- CLI-breaking changes:
  - group submit proposal: takes a single arg which is a proposal.json file

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review. -->

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

amaurym commented on proto/cosmos/group/v1beta1/types.proto:
```diff
@@ -46,23 +46,18 @@ message ThresholdDecisionPolicy {
-  // CHOICE_VETO defines a voting choice with veto.
-  CHOICE_VETO = 4;
+// VoteOption enumerates the valid vote options for a given governance proposal.
+enum VoteOption {
```
Important change 1: rename group's choice to VoteOption (to match gov)

amaurym commented on proto/cosmos/group/v1beta1/tx.proto:
```diff
@@ -226,8 +226,8 @@ enum Exec {
-// MsgCreateProposal is the Msg/CreateProposal request type.
-message MsgCreateProposal {
+// MsgSubmitProposal is the Msg/SubmitProposal request type.
+message MsgSubmitProposal {
```
Important change 2: MsgSubmitProposal everywhere

amaurym commented on proto/cosmos/group/v1beta1/types.proto:
```diff
@@ -162,46 +157,46 @@ message Proposal {
-  // Status defines proposal statuses.
-  enum Status {
+  // ProposalStatus defines proposal statuses.
+  enum ProposalStatus {
```
Important change 3: ProposalStatus everywhee

blushi commented on proto/cosmos/group/v1beta1/types.proto:
```diff
@@ -221,16 +218,16 @@ message Proposal {
 
     // An empty value is not allowed.
-    EXECUTOR_RESULT_UNSPECIFIED = 0 [(gogoproto.enumvalue_customname) = "ProposalExecutorResultInvalid"];
+    EXECUTOR_RESULT_UNSPECIFIED = 0;
```
should also prefix this with `PROPOSAL_`? it's a bit longer but more consistent with the rest

blushi reviewed (commented):
looks good so far, just some small comments

amaurym commented on proto/cosmos/group/v1beta1/types.proto:
```diff
@@ -162,98 +159,99 @@ message Proposal {
-    STATUS_SUBMITTED = 1 [(gogoproto.enumvalue_customname) = "ProposalStatusSubmitted"];
+  // final_tally_result contains the sums of all weighted votes for this
+  // proposal for each vote option, after tallying.
+  TallyResult final_tally_result = 10 [(gogoproto.nullable) = false];
```
Important change 4: Rename Tally to TallyResult to match gov

cmwaters replied:
My very minor problem is that `final_tally_result` isn't necessarily final right? If you query a proposal halfway through the election you're going to get a tally_result that has some votes but not all

amaurym replied:
After yesterday's discussion, we decided to keep `final_tally_result`, and add a tally endpoint to groups #11151

blushi commented on x/group/client/cli/tx.go:
```diff
@@ -476,49 +475,63 @@ func MsgUpdateGroupPolicyMetadataCmd() *cobra.Command {
+Example:
+	$ %s tx gov submit-proposal path/to/proposal.json
+	
+	Where proposal.json contains:
+
+{
+	"group_policy_address": "cosmos1...",
+	// array of proto-JSON-encoded sdk.Msgs
+	"messages": [
+	{
+		"@type": "/cosmos.bank.v1beta1.MsgSend",
+		"from_address": "cosmos1...",
+		"to_address": "cosmos1...",
+		"amount":[{"denom": "stake","amount": "10"}]
+	}
+	],
+	"metadata: "4pIMOgIGx1vZGU=", // base64-encoded metadata
+	"proposers": ["cosmos1...", "cosmos1..."],
+}`, version.AppName),
```
I think `--exec` flag should be kept as a flag, that people don't need to worry about unless they want to use it

cmwaters replied:
What does the `--exec` flag do exactly?

amaurym replied:
it's this field:

https://github.com/cosmos/cosmos-sdk/blob/4addb7368c021617818065a89811ed6e28c621c3/proto/cosmos/group/v1beta1/tx.proto#L251-L254

blushi replied:
indeed, cf. https://github.com/cosmos/cosmos-sdk/issues/11092#issuecomment-1027790198

cmwaters commented on proto/cosmos/group/v1beta1/types.proto:
```diff
@@ -162,98 +159,99 @@ message Proposal {
+  // timeout is the timestamp of the block where the proposal execution times out. Header times of the votes and
+  // execution messages must be before this end time to be included in the election. After the timeout timestamp the
+  // proposal can not be executed anymore and should be considered pending delete.
+  google.protobuf.Timestamp timeout = 11 [(gogoproto.nullable) = false, (gogoproto.stdtime) = true];
```
This is the same as `voting_end_time` within gov right?

amaurym replied:
It seems to me this `timeout` is the max of both `voting_end_time` and execution_end_time

blushi replied:
if we eventually get rid of the ability to execute proposals on demand as described in https://github.com/cosmos/cosmos-sdk/issues/11092#issuecomment-1027790198, then I guess we should eventually rename this to `voting_end_time` too but seems fine to keep it as is for now

amaurym replied:
Since we're not changing the proposal execution model, I propose to keep it as-is.

cmwaters replied:
I think the comment here is still not clear we have two windows: The election window from submission time to the end of voting time where votes are accepted. Then we have the execution window from voting end time to timeout where `MsgExec` can be submitted in the case the proposal passes

amaurym replied:
I don't think we have 2 windows. We just have one window for execution, which is this `timeout`. And obviously voting needs to pass before we do execution, but there's no window for voting only.

I can update the comment.

cmwaters reviewed (commented):
Looks good for the most part. There are some changes which I wonder whether we should change gov to align with groups and not the other way around. We can discuss it at the call later today

amaurym commented on proto/cosmos/group/v1beta1/types.proto:
```diff
@@ -151,8 +148,8 @@ message Proposal {
   // proposers are the account addresses of the proposers.
```
rename `Proposal.proposal_id` to `Proposal.id`

amaurym commented:
I think this is R4R again. Changes since last time:
- rename proposal.proposal_id to id, groupInfo.group_id to id
- add some comments on `final_tally_result`

atheeshp commented on x/group/client/testutil/tx.go:
```diff
@@ -1122,6 +1123,13 @@ func (s *IntegrationTestSuite) TestTxUpdateGroupPolicyMetadata() {
 	}
 }
 
+// TestTxCreateProposal tests submitting proposal.
+//
+// Please don't rename this to TestTxSubmitProposal. It will redefine the order
+// of the tests being run in this file with `go test`, and will mess up the
+// proposal ids in other tests (e.g. voting or Exec tests).
+// This is a headache, but requires a bigger refactor of all tests in this file
+// so that each one is independent.
```
can we create an issue for this then? and can be refactored in upcoming versions.

atheeshp reviewed (approved):
@amaurym can you fix the conflicts, pre-approving.

cmwaters commented on proto/cosmos/group/v1beta1/tx.proto:
```diff
@@ -245,17 +245,17 @@ message MsgCreateProposal {
+  // messages is a list of `sdk.Msg`s that will be executed if the proposal passes.
+  repeated google.protobuf.Any messages = 4;
 
   // exec defines the mode of execution of the proposal,
   // whether it should be executed immediately on creation or not.
   // If so, proposers signatures are considered as Yes votes.
   Exec exec = 5;
```
Just so I understand this correctly, if this is true, the application will create the proposal count up the votes from the proposers and execute the messages if the proposal already passes?

amaurym replied:
exactly.

https://github.com/cosmos/cosmos-sdk/blob/b9c7fd11d356a7fb51d76b8a7f79425fca02e750/x/group/keeper/msg_server.go#L460-L479

cmwaters commented on proto/cosmos/group/v1beta1/types.proto:
```diff
@@ -174,98 +171,101 @@ message Proposal {
-    option (gogoproto.goproto_enum_prefix) = false;
+  // status represents the high level position in the life cycle of the proposal. Initial value is Submitted.
+  ProposalStatus status = 8;
 
-    // An empty value is invalid and not allowed.
-    STATUS_UNSPECIFIED = 0 [(gogoproto.enumvalue_customname) = "ProposalStatusInvalid"];
+  // result is the final result based on the votes and election rule. Initial value is unfinalized.
+  // The result is persisted so that clients can always rely on this state and not have to replicate the logic.
+  ProposalResult result = 9;
```
With `gov` status and result are the same thing right?

amaurym replied:
Yeah. IIUC, we have:
- if status==STATUS_CLOSED, then result==RESULT_ACCEPTED or RESULT_REJECTED
- if status is anything else, then result==RESULT_UNFINALIZED

i'm not sure why the initial design was like this, but IMO I think we could consider removing `STATUS_CLOSED` and `result`, and adding `STATUS_PASSED` and `STATUS_REJECTED` just like gov.

cmwaters commented on proto/cosmos/group/v1beta1/types.proto:
```diff
@@ -174,98 +171,101 @@ message Proposal {
+  // yes is the weighted sum of yes votes.
+  string yes = 1;
 
-  // no_count is the weighted sum of no votes.
-  string no_count = 2;
+  // abstain is the weighted sum of abstainers.
+  string abstain = 2;
 
-  // abstain_count is the weighted sum of abstainers
-  string abstain_count = 3;
+  // no is the weighted sum of no votes.
+  string no = 3;
 
-  // veto_count is the weighted sum of vetoes.
-  string veto_count = 4;
+  // no_with_veto is the weighted sum of veto.
+  string no_with_veto = 4;
 }
```
quick poll: what do we think is more intuitive: "Yes" or "YesCount" etc... ?

amaurym replied:
Hmm, the fact that it's a string can be confusing, I prefer `_count` so that we know it represents a number. Will change in gov too.

cmwaters reviewed (approved):
I've left a few more comments but this mostly looks good to me and is ready to merge

amaurym commented:
All comments addressed, some might need a bigger decision. I'm putting automerge on.

marbar3778 commented:
what is reasoning behind breaking v1beta1 messages? it would be nice to not break the old api and only change the new one

marbar3778 reviewed (commented):
this sort of v1beta1 message breakage would usually constitute a version bump. Would this not cause  a headache for integrators. They now have to migrate to new message types then migrate to message passing.

amaurym commented:
Which v1beta1 messages are broken? This PR just changes gov@v1beta2 and group@v1beta1 stuff.
